### PR TITLE
fix: SSH Tunnel creation with dynamic form

### PR DESCRIPTION
**Pull request by hughhhh**
### SUMMARY

Fixing issue with SSH tunnel creation with dynamic forms. Main issue was the ssh toggle would lose state, and validate_parameters endpoints would fail immediately since no tunnel is created on every `onBlur` event

To fix it I refactored the toggle state to the reducer, and stop validation on any ssh tunnel requests to connect to a DB

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/features/databases/DatabaseModal/index.tsx [hunk 1]
```diff
@@ -763,15 +763,19 @@ const DatabaseModal: FunctionComponent<DatabaseModalProps> = ({
         });
       }
 
-      // make sure that button spinner animates
-      setLoading(true);
-      const errors = await getValidation(dbToUpdate, true);
-      if ((validationErrors && !isEmpty(validationErrors)) || errors) {
+      // only do validation for non ssh tunnel connections
+      console.log(dbToUpdate);
```
Comment (Antonio-RiveroMartnez) [inline superset-frontend/src/features/databases/DatabaseModal/index.tsx:767]: leftover

File: superset-frontend/src/features/databases/DatabaseModal/DatabaseConnectionForm/CommonParameters.tsx [hunk 2]
```diff
@@ -250,3 +251,34 @@ export const forceSSLField = ({
     />
   </div>
 );
+
+export const SSHTunnelSwitch = ({
+  isEditMode,
+  changeMethods,
+  db,
+}: FieldPropTypes) =>
+  true ? (
```
Comment (betodealmeida) [inline superset-frontend/src/features/databases/DatabaseModal/DatabaseConnectionForm/CommonParameters.tsx:260]: Why the ternary operator here if the condition is always true?

File: superset-frontend/src/views/CRUD/hooks.ts [hunk 3]
```diff
@@ -753,6 +753,11 @@ export function useDatabaseValidation() {
         .catch(e => {
           if (typeof e.json === 'function') {
             return e.json().then(({ errors = [] }: JsonObject) => {
+              if (database?.parameters?.ssh) {
```
Comment (betodealmeida) [inline superset-frontend/src/views/CRUD/hooks.ts:756]: I don't think this belongs here... if we don't want to show validation errors for some reason, then we shouldn't be running the validation API calls.

**Review by betodealmeida — CHANGES_REQUESTED** [review]

File: superset-frontend/src/features/databases/DatabaseModal/index.tsx [hunk 4]
```diff
@@ -763,15 +763,18 @@ const DatabaseModal: FunctionComponent<DatabaseModalProps> = ({
         });
       }
 
-      // make sure that button spinner animates
-      setLoading(true);
-      const errors = await getValidation(dbToUpdate, true);
-      if ((validationErrors && !isEmpty(validationErrors)) || errors) {
+      // only do validation for non ssh tunnel connections
+      if (!dbToUpdate?.ssh_tunnel) {
```
[the remark this answers is not in the document]
Comment (hughhhh) [reply superset-frontend/src/features/databases/DatabaseModal/index.tsx:767]: for the initial implementation, we were okay with doing validation on just connect. Right now users are not able to connect at all via dynamic forms.

I'm open to this fix but that's going to take longer to implement vs. fixing the issue at hand.

Let me know what you think

**Review by betodealmeida — APPROVED** [review]
Awesome!

**Comment by sadpandajoe** [comment]
🏷️ preset:2023.27

**Pull request merged**